crates/wasi/src/preview2/host/filesystem.rs

```diff
@@ -15,16 +19,26 @@ impl From<TableError> for types::Error {
 }
 
 impl<T: WasiView> preopens::Host for T {
-    fn get_directories(&mut self) -> Result<Vec<(types::Descriptor, String)>, anyhow::Error> {
-        Ok(self.ctx().preopens.clone())
+    fn get_directories(
+        &mut self,
+    ) -> Result<Vec<(Resource<types::Descriptor>, String)>, anyhow::Error> {
+        Ok(self
+            .ctx()
+            .preopens
+            .iter()
+            .map(|(fd, name)| (Resource::new_borrow(fd.rep()), name.clone()))
```

Comment:
This is I believe going to return an error at runtime because borrows can't be returned, they can only be taken.

Two ways I can think of to handle this are:

* Don't store preopens as `Resource<Descriptor>`. Instead insert them into the table every time `get-directories` is called.
* Alternatively on `HostDescriptor::drop` don't actually delete preopens from the table. This will "leak" them in the table but that's ok since it's a static set that doesn't grow.

Reply:
I made it insert them into the table every tine `get-directories` is called.